### PR TITLE
LibC: Fix sigsetjmp and siglongjmp

### PR DESCRIPTION
**Pull request by boricj**
`sigsetjmp()` and `siglongjmp()` have a critical bug that hijack the execution flow of their caller by returning to the wrong call site. This breaks notably `mksh` by exhibiting erratic behavior during usage. As now written in `setjmp.h`:
```c++
/**
 * Calling conventions mandates that siglongjmp() cannot call setjmp(),
 * otherwise the restored calling environment will not be the original caller's
 * but siglongjmp()'s and we'll return to the wrong call site.
 *
 * The setjmp(), sigsetjmp() and longjmp() functions have to be implemented in
 * assembly because they touch the call stack and registers in non-portable
 * ways. However, we *can* implement siglongjmp() as a standard C function.
 */
```

### CODE REVIEW

**Comment by boricj** [comment]
I've tested the assembly bits only on i686, but it _seems_ to work with bash and mksh on that port.

File: Userland/Libraries/LibC/arch/i386/setjmp.S [hunk 1]
```diff
@@ -4,34 +4,63 @@
...
-    mov 4(%esp), %eax
-    mov %ebx, 0(%eax)
-    mov %esi, 4(%eax)
-    mov %edi, 8(%eax)
-    mov %ebp, 12(%eax)
-    lea 4(%esp), %ecx
-    mov %ecx, 16(%eax)
-    mov (%esp), %ecx
-    mov %ecx, 20(%eax)
+    movl $0, 8(%esp)        // Set val argument to 0
```
Comment (mirabilos) [inline Userland/Libraries/LibC/arch/i386/setjmp.S:15]:
```
01:12⎜<RT|AO:#!/bin/mksh> SerenityOS side decided to fix sig*jmp, thats great.
     ⎜    https://github.com/SerenityOS/serenity/pull/9392/commits/535cc24e1b70c8e985d8fa3edad1ff5b5f1f67fe
01:17⎜<mirabilos:#!/bin/mksh> nice
01:20⎜<mirabilos:#!/bin/mksh> https://github.com/SerenityOS/serenity/blob/535cc24e1b70c8e985d8fa3edad1ff5b5f1f
     ⎜    67fe/Userland/Libraries/LibC/arch/i386/setjmp.S#L15
01:21⎜<mirabilos:#!/bin/mksh> congratulations, they have just overflowed the stack
01:21⎜<mirabilos:#!/bin/mksh> I mean, it’s a nice idea in theory
01:21⎜<mirabilos:#!/bin/mksh> but setjmp() is called with one argument only
01:21⎜<mirabilos:#!/bin/mksh> they try to transform setjmp(jb) to sigsetjmp(jb,0)
01:21⎜<mirabilos:#!/bin/mksh> but there’s nothing on stack reserved to store that 0
01:21⎜<mirabilos:#!/bin/mksh> better do:
01:22⎜<mirabilos:#!/bin/mksh> mov ecx,4[esp]
01:22⎜<mirabilos:#!/bin/mksh> xor eax,eax
01:22⎜<mirabilos:#!/bin/mksh> jmp 1f
01:22⎜<mirabilos:#!/bin/mksh> sigsetjmp:
01:22⎜<mirabilos:#!/bin/mksh> mov ecx,4[esp]
01:22⎜<mirabilos:#!/bin/mksh> mov eax,8[esp]
01:22⎜<mirabilos:#!/bin/mksh> 1:
```

File: Userland/Libraries/LibC/setjmp.h [hunk 2]
```diff
@@ -14,25 +14,51 @@
...
-    bool did_save_signal_mask;
+    int did_save_signal_mask;
     sigset_t saved_signal_mask;
 };
 
 typedef struct __jmp_buf jmp_buf[1];
 typedef struct __jmp_buf sigjmp_buf[1];
 
+/**
+ * Calling conventions mandates that siglongjmp() cannot call setjmp(),
```
Comment (mirabilos) [inline Userland/Libraries/LibC/setjmp.h:49]: sigsetjmp?

File: Userland/Libraries/LibC/arch/i386/setjmp.S (hunk 1 above)
Comment (boricj) [reply Userland/Libraries/LibC/arch/i386/setjmp.S]: I guess that shows I'm not proficient at all in x86 assembly, especially in tricky situations like this... I've amended the PR with what I hope is a correct fix.

Comment (mirabilos) [reply Userland/Libraries/LibC/arch/i386/setjmp.S]: This looks-ish better (I don’t normally do AT&T syntax), thanks.

**Pull request merged**